Feature: Video to Audio to Effect - #1 - Get Transcript and Toxicity Score of Transcript

rwackes opened the pull request:
In the first 9 commits: 

- I updated the HTML/CSS for the initial webpage (a webpage that introduces users to our product and our team and directs users to use our product), the upload webpage (a webpage where users can upload the video they wish to be reviewed), and the results page (a webpage that displays the results/effects of our analysis done on the video). 

- I implemented Cloud Video Intelligence API that takes in the path of a video file stored in a Google Cloud Storage Bucket to generate a speech transcription of the video. When a user uploads their video file and navigates to the results webpage, the transcription begins being generated. Once found, the most likely transcription is displayed on the webpage to the user for now. This transcription will be used in the process of determining the effect and potential toxicity of the video's audio. 

In the next 5 commits:

- I implemented the Perspective API that takes the transcription in plain text and builds an appropriate AnalyzeCommentRequest according to the PerspectiveAPI guidelines and returns an AnalyzeCommentResponse with the determined "toxicity" score of the transcription. Toxicity scores range [0,1] with the higher the score, the likelier chance the transcription will be perceived as "toxic."

- I used Guava Java-based library developed by Google to write concurrent code around Perspective API with ListenableFuture. ListenableFuture extends Future interface of JDK to represent the result of an asynchronous computation (which in this case is the Perspective API call). ListenableFuture registers when the computation is complete and when callbacks can be executed, which is important in supporting our operation to get the toxicity score returned within the AnalyzeCommentResponse.

- I used Fasterxml Jackson as a JSON parser for Java in order to serialize our Java objects to JSON for our AnalyzeCommentRequest being sent to the Perspective API and to serialize the JSON response that the Perspective API returned to us.  

- I used the AsyncHttpClient library to asynchronously execute HTTP request to and process the response from Perspective API. This is used with ListenableFuture so that the listener will run when the HTTP call is completed.

sharond106 commented on capstone/src/main/java/com/google/sps/perspective/request/BaseRequest.java:
```diff
@@ -0,0 +1,66 @@
+import com.google.common.util.concurrent.MoreExecutors;
+import org.asynchttpclient.Response;
+import org.asynchttpclient.extras.guava.ListenableFutureAdapter;
+import javax.annotation.Nullable;
+
+abstract class BaseRequest<R> {
+
```
add a comment here for BaseRequest class

rwackes replied:
Done

sharond106 commented on capstone/src/main/java/com/google/sps/perspective/request/BaseRequest.java:
```diff
@@ -0,0 +1,66 @@
+  BaseRequest(Client client) {
+    this.client = client;
+  }
+
+  public ListenableFuture<R> postAsync() {
+    String body = null;
```
could you add a comment for this function?

rwackes replied:
Done

sharond106 commented on capstone/src/main/java/com/google/sps/perspective/request/Client.java:
```diff
@@ -0,0 +1,46 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.asynchttpclient.AsyncHttpClient;
+import org.asynchttpclient.DefaultAsyncHttpClient;
+
+public class Client {
+
```
add a comment

rwackes replied:
Done

sharond106 commented on capstone/src/main/java/com/google/sps/perspective/response/AnalyzeCommentResponse.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class AnalyzeCommentResponse extends BaseResponse {
+
```
add a comment :)

rwackes replied:
Done :)

sharond106 commented on capstone/src/main/java/com/google/sps/servlets/AudioEffectServlet.java:
```diff
@@ -0,0 +1,89 @@
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+@WebServlet("/audio-effect")
```
update comment

rwackes replied:
Done

sharond106 commented:
I agree with Cassandra! I just marked a few more places where comments would be helpful, but once you finish updating the code, I'll try to understand how the functions are working more!

rwackes commented on capstone/src/main/java/com/google/sps/data/Analyze.java:
```diff
@@ -0,0 +1,131 @@
+    String transcription = "";
+
+    // Instantiate Video Intelligence.
+    try (VideoIntelligenceServiceClient client = VideoIntelligenceServiceClient.create()) {
```
This is a try-with-resources statement. What this means is that the code inside the try parameter which instantiates a client for video intelligence API usage is automatically closed once the code executes or if something breaks unexpectedly and the code inside the try does not execute completely, but the VI API instantiation will be closed still. I am catching any exceptions it might throw in the smaller method above this one.

rwackes commented on capstone/src/main/java/com/google/sps/data/Analyze.java:
```diff
@@ -0,0 +1,131 @@
+      VideoContext context = VideoContext.newBuilder().setSpeechTranscriptionConfig(config).build();
+
+      // Create the request. 
+      AnnotateVideoRequest request = AnnotateVideoRequest.newBuilder().setInputUri(gcsUri).addFeatures(Feature.SPEECH_TRANSCRIPTION).setVideoContext(context).build();
```
Done

rwackes commented on capstone/src/main/webapp/index.html:
```diff
@@ -1,14 +1,119 @@
+    </section>
+
+    <!-- JQuery library, Popper JS, and Bootstrap JS -->
+    <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
```
I put it there because Bootstrap website says: "Place the following <script>s near the end of your pages, right before the closing </body> tag, to enable them. jQuery must come first, then Popper.js, and then our JavaScript plugins."

rwackes commented on capstone/src/main/webapp/js/extract-audio-script.js:
```diff
@@ -12,3 +12,30 @@
+  // Fetch the audio transcription of the passed in video.
+  fetch('/audio-effect', {
+    method: 'GET'
+  }).then(response => response.text()).then((effect) => {
```
Done

rwackes commented on capstone/src/main/webapp/upload.html:
```diff
@@ -0,0 +1,61 @@
+    </div>
+
+  <!-- JQuery library, Popper JS, and Bootstrap JS -->
+    <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
```
See response to other same comment

rwackes commented on capstone/src/main/java/com/google/sps/data/Analyze.java:
```diff
@@ -0,0 +1,131 @@
+
+  /**
+   * Set the gcsUri for video file being analyzed.
+   * @return a string containing the audio transcription of the video file
```
Thanks! :)

sharond106 commented on capstone/src/main/java/com/google/sps/perspective/PerspectiveAPIBuilder.java:
```diff
@@ -0,0 +1,77 @@
+   * @param sessionId an opaque session id
+   * @return the builder
+   */
+  public PerspectiveAPIBuilder setSessionId(String sessionId) {
+    this.sessionId = sessionId;
+    return this;
+  }
+
```
I dont't think you're calling this anywhere?_

rwackes replied:
My thought was that I would need this for unit testing. I am working on that currently so when I'm done, if I find that this was unnecessary I will go ahead and edit these files.

wdi-gg replied:
Nice planning ahead, and I would suggest:
1. Only add a function when it's actually used (do not over design for future usage / testing).
2. When adding tests, put methods created for testing together and use @VisibleForTesting annotation for them.

Same below.

sharond106 commented on capstone/src/main/java/com/google/sps/perspective/PerspectiveAPIBuilder.java:
```diff
@@ -0,0 +1,77 @@
+   * @param clientToken an opaque token that is echoed back in the response
+   * @return the builder
+   */
+  public PerspectiveAPIBuilder setClientToken(String clientToken) {
+    this.clientToken = clientToken;
+    return this;
+  }
+
```
Same with this one

rwackes replied:
Same answer as above

sharond106 commented on capstone/src/main/java/com/google/sps/perspective/PerspectiveAPIBuilder.java:
```diff
@@ -0,0 +1,77 @@
+   * @param apiVersion the api version for Perspective API
+   * @return the builder
+   */
+  public PerspectiveAPIBuilder setApiVersion(String apiVersion) {
+    this.apiVersion = apiVersion;
+    return this;
+  }
+
```
Or this one. I'm just doing control+f, and none of these are showing up, so just check if you need them!

rwackes replied:
Same as answer as above

sharond106 commented on capstone/src/main/java/com/google/sps/perspective/attributes/Attribute.java:
```diff
@@ -0,0 +1,94 @@
+   * @param scoreType the type of the score
+   * @return the builder
+   */
+  public Attribute setScoreType(String scoreType) {
+    this.scoreType = scoreType;
+    return this;
+  }
+  
+  /**
+   * Sets the threshold that scores must be at or above in order to be returned.
+   * @param scoreThreshold the threshold for scores to return
+   * @return the builder
+   */
+  public Attribute setScoreThreshold(float scoreThreshold) {
+    this.scoreThreshold = scoreThreshold;
+    return this;
+  }
+}
```
I dont think these 2 functions are being used

rwackes replied:
Same as above for setScoreType (meant to be set as PLAIN_TEXT). setScoreThreshold is not as I decided to display scores through a meter regardless of their value. I will update accordingly.

sharond106 commented on capstone/src/main/java/com/google/sps/perspective/request/AnalyzeCommentRequest.java:
```diff
@@ -0,0 +1,123 @@
+   * @param type the type of the text of the comment
+   * @return the AnalyzeCommentRequest
+   */
+  public AnalyzeCommentRequest setComment(String comment, String type) {
+    requestBody.comment = new Comment(comment, type);
+    return this;
+  }
+
+  /**
+   * Creates a new instance of default Comment.
+   * @param comment the text of the comment
+   * @return the AnalyzeCommentRequest
+   */
+  public AnalyzeCommentRequest setComment(String comment) {
+    return setComment(comment, null);
+  }
```
It seems like you are only ever calling the second setComment() method, so maybe you can condense these 2 if you never plan on using the other one.

rwackes replied:
Same as above

sharond106 commented on capstone/src/main/java/com/google/sps/perspective/request/BaseRequest.java:
```diff
@@ -0,0 +1,83 @@
+    String body = null;
+    try {
+      body = bodyJSON();
+    } catch (JsonProcessingException e) {
+      // TODO: Handle accordingly
+    }
+
```
I don't totally understand how this part is working, but maybe just set body = "" ? It seems like you need the body variable to return a ListenableFuture

rwackes replied:
I don't want to set the body to an actual string and I want to leave it as null so that my try/catch block in my servlet can catch that error when it is thrown and I can return a simple message explaining the error to the user through my JS

sharond106 commented on capstone/src/main/java/com/google/sps/perspective/response/AttributeScores.java:
```diff
@@ -0,0 +1,35 @@
+ * hence spanScores.
+ * A comment is given an overall score for the entire comment for a given attribute, hence
+ * summaryScore.
+ */
+public class AttributeScores {
+
+  /**
+   * Note: spanScores would be another JsonProperty here, but our team does not care about
+   * span scores in our Video Vigilance implementation.
+   */
+
+  @JsonProperty("summaryScore")
+  public ProbabilityScore summaryScore;
+}
```
What is this class's purpose? I can see that you're using it in the AnalyzeCommentResponse class, but when is it ever set? It seems like it would always be null and have a size of 0 in getAttributeSummaryScores().

rwackes replied:
This class much like some others is meant to interpret the JSON response of Perspective API, which I serialize to Java. Attribute summary scores are set when the Perspective API call executes and completes and AnalyzeCommentResponse is returned. When it returns, I serialize it from a JsonProperty to a Java instance variable. The map and the float values are set and necessary for when I call getAttributeSummaryScore() in my java servlet that interacts with VI API and Perspective API and returns a float value corresponding to the passed in attribute type.

cassmarcussen commented on capstone/src/main/java/com/google/sps/perspective/request/Client.java:
```diff
@@ -0,0 +1,54 @@
+/**
+ * Configure the client for Perspective API. 
+ */
+public class Client {
```
Could you give this a more descriptive name?

wdi-gg commented on capstone/src/main/java/com/google/sps/perspective/attributes/Comment.java:
```diff
@@ -0,0 +1,39 @@
+  /**
+   * Creates an instance of a comment to be analyzed by Perspective API.
+   * @param text is the comment text
+   * @param type is the type of the text for the comment
```
What are the types, TOXICITY/INSULT etc.?

rwackes replied:
There are different types for different things in Perspective. In regards to a comment, the type is actually PLAIN_TEXT. A comment can be of type PLAIN_TEXT or HTML. In regards to an attribute, the type is TOXICITY/INSULT/etc. In regards to a score, the type is PROBABILITY.

rwackes replied:
Updated comment for param tag to reflect explanation. Done!

wdi-gg commented on capstone/src/main/java/com/google/sps/servlets/AudioEffectServlet.java:
```diff
@@ -0,0 +1,89 @@
+package com.google.sps.servlets;
+
+import com.google.gson.Gson;
+/**
```
Clean up?

rwackes replied:
Done!

wdi-gg commented on capstone/src/main/java/com/google/sps/servlets/AudioEffectServlet.java:
```diff
@@ -0,0 +1,89 @@
+      // Create an AnalyzeCommentRequest for the transcription and store the response.
+      ListenableFuture<AnalyzeCommentResponse> future = api.analyze()
+        .setComment(audioTranscription)
+        .addAttribute(Attribute.ofType(Attribute.TOXICITY))
```
Why are we only adding TOXICITY but not other types?

rwackes replied:
This was just for my initial implementation to ensure that my connection between VI API and Perspective API worked. I took this further in future PRs to add more attributes to score the comment on and to handle the multiple scores/how we display to the user.

vkkpng-g commented on capstone/src/main/java/com/google/sps/data/Analyze.java:
```diff
@@ -0,0 +1,143 @@
+
+
+/**
+ * One java file that centralizes the methods of Cloud Video Intelligence.
```
Can you write  what kind of methods to explicitly state it out?

rwackes replied:
My idea was that Sharon and I would have our methods using Video Intelligence API in one file to "centralize" all the VI API methods, but in order to promote better collaboration this idea was not followed through with. To avoid confusion, I've now removed this comment. Done!

vkkpng-g commented on capstone/src/main/java/com/google/sps/data/Analyze.java:
```diff
@@ -0,0 +1,143 @@
+        client.annotateVideoAsync(request);
+
+      // Grab the transcription generated.
+      for (VideoAnnotationResults result : response.get(600, TimeUnit.SECONDS).getAnnotationResultsList()) {
```
Instead of nested for loop can we do

VideoAnnotationResults result  = response.get(600, TimeUnit.SECONDS).getAnnotationResultsList()
for (SpeechTranscription speechTranscription : result.getSpeechTranscriptionsList()) 

Then we get rid of the for loop

rwackes replied:
Done, though instead of getAnnotationResultsList() I used getAnnotationResults(0) to get the first (and only one) so there is no list to loop through.

vkkpng-g commented on capstone/src/main/java/com/google/sps/data/Analyze.java:
```diff
@@ -0,0 +1,143 @@
+        if (result.getSpeechTranscriptionsList().size() == 0) {
+          continue;
+        }
+        for (SpeechTranscription speechTranscription : result.getSpeechTranscriptionsList()) {
```
if result.getSpeechTranscriptionsList() returns 2  speechTranscription

[1] the first one has 0 getAlternativesCount()
[2] the second one has 5 getAlternativesCount()

Doesn't that mean your return statement will be

"Hardcoded message. If this returns, that means there was no transcription found. Transcription Text.Transcription Text.Transcription Text.Transcription Text.Transcription Text."

vkkpng-g commented on capstone/src/main/java/com/google/sps/servlets/AudioEffectServlet.java:
```diff
@@ -0,0 +1,89 @@
+    try {
+      // Instantiate and build the PerspectiveAPIBuilder with my API key.
+      PerspectiveAPI api = new PerspectiveAPIBuilder()
+        .setApiKey("AIzaSyCx72YUXfGl2npdgwyY8ZLXLNAc-vgks7w")
```
lets move the apikey as a const/global

rwackes commented on capstone/src/main/java/com/google/sps/data/Analyze.java:
```diff
@@ -0,0 +1,143 @@
+   * Set the gcsUri for video file being analyzed.
+   * @return a string containing the audio transcription of the video file
+   */
+  public static String transcribeAudio() {
```
I use this method to try what is inside the transcribeAudio(String gcsUri) method. If anything were to go wrong when setting the annotatevideorequest/speechtranscriptionconfiguration/etc or when attempting to go through the speechtranscriptionlist in transcribeAudio(String gcsUri) method, then transcribeAudio() method (without the param) would catch the errors/exceptions thrown.

wdi-gg replied:
It seems try-with-resources statement might also be followed with a catch statement to catch exceptions, so that we don't need an extra layer to do that.
More context: https://docs.oracle.com/javase/specs/jls/se7/html/jls-14.html#jls-14.20.3.2

rwackes commented on capstone/src/main/java/com/google/sps/data/Analyze.java:
```diff
@@ -0,0 +1,143 @@
+      String gcsUri = "gs://video-vigilance-videos/youtube_ad_test_2.mp4";
+      return transcribeAudio(gcsUri);
+    } catch (Exception e) {
+      System.out.println("Exception while running:\n" + e.getMessage() + "\n");
```
Done!

wdi-gg replied:
Can we remove all debug statements through out the file?

rwackes commented on capstone/src/main/java/com/google/sps/data/Analyze.java:
```diff
@@ -0,0 +1,143 @@
+        client.annotateVideoAsync(request);
+
+      // Grab the transcription generated.
+      for (VideoAnnotationResults result : response.get(600, TimeUnit.SECONDS).getAnnotationResultsList()) {
```
It's used to wait for the video to be processed before moving forward in the code. Done!

rwackes commented on capstone/src/main/java/com/google/sps/data/Analyze.java:
```diff
@@ -0,0 +1,143 @@
+
+      // Grab the transcription generated.
+      for (VideoAnnotationResults result : response.get(600, TimeUnit.SECONDS).getAnnotationResultsList()) {
+        if (result.getSpeechTranscriptionsList().size() == 0) {
```
Used for debugging. Removed. Done!

rwackes commented on capstone/src/main/java/com/google/sps/data/Analyze.java:
```diff
@@ -0,0 +1,143 @@
+              SpeechRecognitionAlternative alternative = speechTranscription.getAlternatives(0);
+              transcription = transcription + alternative.getTranscript();
+
+              /**
```
I kept it in in case we had enough time and could implement this as a stretch goal. FE, in my second PR, I implemented confidence. I deleted this comment though so it is not present when we merge to master. Done!

rwackes commented on capstone/src/main/java/com/google/sps/data/Analyze.java:
```diff
@@ -0,0 +1,143 @@
+            if (speechTranscription.getAlternativesCount() > 0) {
+              // Get the most likely transcription.
+              SpeechRecognitionAlternative alternative = speechTranscription.getAlternatives(0);
+              transcription = transcription + alternative.getTranscript();
```
The outer for loop is used to get the first result from the results list. Since we are sending in one video to be processed, there should only be one result in the list. 

The inner for loop is used to loop through the segments of the transcription, which are stored in a list. The way Video Intelligence API breaks down and returns a transcription for a video is that it breaks the video up into "spans" or segments and generates transcriptions for each segment of the video. So, the inner for loop actually loops through each segment of the video's transcription and gets the most likely transcription generated for each segment. 

I deleted the outer for loop and added more comments to reflect this explanation more clearly in my code. Done!

rwackes commented:
Screenshot of diagram explaining classes' functionality and interaction for this PR: https://screenshot.googleplex.com/n6uNzt8uALt

wdi-gg commented on capstone/src/main/java/com/google/sps/perspective/request/AnalyzeCommentRequest.java:
```diff
@@ -0,0 +1,123 @@
+   * @return the AnalyzeCommentRequest
+   */
+  public AnalyzeCommentRequest setComment(String comment) {
+    return setComment(comment, null);
```
I think you mentioned type is generally PLAIN_TEXT. Is there a reason we set it to null?

rwackes replied:
Since only PLAIN_TEXT is currently supported (see comment.type field and description: https://github.com/conversationai/perspectiveapi/blob/master/2-api/methods.md#analyzecomment-request ), setting the type to null made the Perspective API set the comment type to PLAIN_TEXT anyways. But because it's pretty much the same and to remove any confusion, I have now changed null to "PLAIN_TEXT". Done!

wdi-gg commented:
> Screenshot of diagram explaining classes' functionality and interaction for this PR: https://screenshot.googleplex.com/XvBki9wwfOS

Thanks for taking time to build this diagram! This is super helpful for understanding the flow.
From reading the design doc and comparing with the flow, I would like to list my understanding and make some high level questions and suggestions:
0. AudioEffectServlet.java contains the main logic.
1. Video to Transcript:
Analyze.java is performing this layer. [Suggestion] Maybe rename it to Transcribe.java or something to avoid using the same word with later layers.
2. Transcript to Effect
2.1 PerspectiveAPIBuilder.java
It's nice to practice using builder, but [Question] I'm not sure if it's actually necessary for our simple case? More context: https://screenshot.googleplex.com/AXKEoY3mfgy
2.2 AnalyzeCommentBody.java, Comment.java and Attribute.java
[Question] Does Perspective API have a doc about what fields are required in a request? Let's add a link to the code.
[Question] If these fields / classes are required, I think the API should provide a package that you can import and use directly, rather than have each user write all these classes on their own.
2.3 AnalyzeCommentResponse.java
[Question] the same question as for request, shouldn't this be provided by the API?
Another [question]: What is AudioTranscriptionServlet.java and where is it used?

vkkpng-g commented on capstone/src/main/java/com/google/sps/perspective/PerspectiveAPIBuilder.java:
```diff
@@ -0,0 +1,77 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.perspective;
```
Is there a particular reason we want to separate the code from PerspectiveAPI?

rwackes replied:
I chose to divide them up because I chose to use a builder pattern. Through the builder pattern, I separate the construction (PerspectiveAPIBuilder) from the representation (PerspectiveAPI). The construction is done in the PerspectiveAPIBuilder class itself. The representation, PerspectiveAPI, is what we see as the user of the class. By separating these two, the representation of PerspectiveAPI class is a lot simpler, neater and readable.

When trying to combine the two and use a factory pattern, such as create(), I ran into many non-static / static context errors due to the fact that I am creating an instance of Client (which is non-static) in my PerspectiveAPI representation after building it through PerspectiveAPIBuilder. If I were to combine PerspectiveAPI and PerspectiveAPIBuilder into one and use create(), I could not make a call to Client to set the endpoint for our AnalyzeComment request.

vkkpng-g replied:
got it, thanks for the explanation. Leaving it open so Di can read this too.

Di feel free to close this comment once u ack her comments.

vkkpng-g commented on capstone/src/main/java/com/google/sps/perspective/response/AnalyzeCommentResponse.java:
```diff
@@ -0,0 +1,67 @@
+   * @return a mapping of Attribute names to the summary score as a probability
+   */
+  public Map<String, Float> getAttributeSummaryScores() {
+    if (attributeSummaryScores == null) {
```
remove the nested if else

if (attributeSummaryScores == null && attributeScores == null) {
} else {
}

rwackes replied:
Done!

vkkpng-g commented on capstone/src/main/java/com/google/sps/perspective/response/BaseResponse.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * The base response for an AnalyzeCommentResponse. 
+ */
+public class BaseResponse {
```
Since BaseResponse is only used once why not combine this file with AnalyzeCommentResponse.

If we do have more time in this project, did u intend to have multiple classes extend BaseResponse? Is that the reason why we have this BaseResponse class?

rwackes replied:
Combined in my second pull request, PR #8, since that is my PR for refactoring/optimizing code. Done.

rwackes commented:
I did not mean to close the PR. Ignore that mistake.

rwackes commented:
In response to Di's comments and questions,
1. I have renamed this file to reflect its behavior more clearly. I had this original idea that both Sharon and I would have our Cloud Video Intelligence method calls (detect shots and transcribe audio methods) in one file, and so I gave it a general name. However, in order to support collaboration in different branches and the single responsibility principle, we did not follow through with this idea. Now, this class is dedicated solely to transcribing the video's audio and the comments/annotations and file name reflect the intention better.
2.1 I would definitely like to use Builder when creating the AnalyzeCommentRequest because of the number of fields, but I can definitely try switching to ".create()" instead of builder when creating the Perspective API call through which we send the built out AnalyzeCommentRequest. I think this is what you were suggesting, but correct me if I am wrong. :)
2.2 Link to methods/fields for a request: https://github.com/conversationai/perspectiveapi/blob/master/2-api/methods.md#scoring-comments-analyzecomment. Will add several links such as this one throughout my code.
2.2 Perspective API has packages for Python, Node.js, and cURL for easy use of their API (see here: https://github.com/conversationai/perspectiveapi/blob/master/1-get-started/sample.md). However, there is no official supported Java package, which is why I had to create all these classes to wrap around the logic of the API myself for our project. 
2.3 Thanks for pointing this out. My bad. I was not using AudioTranscriptionServlet.java anymore. In my very first implementation, I had one servlet dedicated to getting the transcription from Video Intelligence API and another dedicated to getting the scores from Perspective API, but I ran into problems trying to pass in the transcription into the second servlet as a parameter, so I combined the two into AudioEffectServlet.java, rendering AudioTranscriptionServlet.java useless. I deleted AudioTranscriptionServlet.java.

wdi-gg commented on capstone/src/main/java/com/google/sps/perspective/request/AnalyzeCommentRequest.java:
```diff
@@ -0,0 +1,122 @@
+   * @return the AnalyzeCommentRequest
+   */
+  public AnalyzeCommentRequest setComment(String comment) {
+    return setComment(comment, "PLAIN_TEXT");
```
Add a comment for "PLAIN_TEXT".

rwackes replied:
Done

wdi-gg commented on capstone/src/main/java/com/google/sps/perspective/response/AnalyzeCommentResponse.java:
```diff
@@ -0,0 +1,67 @@
+   * @param attr Attribute Name
+   * @return the summary score as a probability
+   */
+  public float getAttributeSummaryScore(String attr) {
```
Maybe combine the methods since it's just one line?

rwackes replied:
The method at line 45 can be used for only returning the summary score of ONE specific attribute whereas the method at line 53 is for returning a Map with the summary score of ALL attributes that were set in the request. So if in my request I specified toxicity, insult, and profanity, but I wanted easy access to the summary score of toxicity in just one line, I would use this method. If in my request I specified toxicity, insult, and profanity, and I wanted easy access to all the summary scores, I would use the second method. I can't add this line from the first method to the second method because that is not the functionality I want in the second method.

wdi-gg commented on capstone/src/main/java/com/google/sps/perspective/response/AnalyzeCommentResponse.java:
```diff
@@ -0,0 +1,67 @@
+  private Map<String, Float> attributeSummaryScores;
+
+  /**
+   * Returns the summary score as a probability.
```
No need to change but it seems the description and @return are always repetitive?

vkkpng-g commented on capstone/src/main/webapp/js/extract-audio-script.js:
```diff
@@ -12,3 +12,30 @@
+window.onload = function() { createAudioTranscription() };
+
+function createAudioTranscription() {
+  console.log('Fetching audio transcription and effect.');
```
is this intentional/i don't recall using these for the demo. (we use the circular thing?)

rwackes replied:
If you're referring to the console.log() statements, they print out in my JS console (which I can view by inspecting the deployed webpage), and this is very helpful for me (the developer) so I could figure at what point in the request our application is at at a certain time and if anything went wrong, where the application was last at. This is not being printed to the user on the webpage, it is being printed to the JS console, much like System.out.println() lines in Java files print to the terminal. However, noting earlier comment to remove debug statement, I'll remove all console.log() lines as well.

rwackes replied:
Removed. BTW: loader is not implemented in this PR since this is the very first PR/branch and was not for UI

vkkpng-g commented on capstone/src/main/webapp/js/extract-audio-script.js:
```diff
@@ -12,3 +12,30 @@
+  fetch('/audio-effect', {
+    method: 'GET'
+  }).then(response => response.text()).then((effect) => {
+    console.log('Fetched audio transcription and effect: ' + effect);
```
Remove if we're not showing to user

rwackes replied:
Done

vkkpng-g commented on capstone/src/main/java/com/google/sps/data/Transcribe.java:
```diff
@@ -0,0 +1,122 @@
+            SpeechRecognitionAlternative alternative = speechTranscription.getAlternatives(0);
+            transcription = transcription + alternative.getTranscript();
+          } else {
+            System.out.println("No transcription found");
```
Remove

rwackes replied:
Done

vkkpng-g commented on capstone/src/main/java/com/google/sps/data/Transcribe.java:
```diff
@@ -0,0 +1,122 @@
+            transcription = "Hardcoded message. If this returns, that means there was no transcription found.";
+          }
+        } catch (IndexOutOfBoundsException ioe) {
+          System.out.println("Could not retrieve frame: " + ioe.getMessage());
```
Remove?

rwackes replied:
Done

vkkpng-g commented on capstone/src/main/java/com/google/sps/perspective/PerspectiveAPIBuilder.java:
```diff
@@ -0,0 +1,56 @@
+   * @param apiVersion the api version for Perspective API
+   * @return the builder
+   */
+  public PerspectiveAPIBuilder setApiVersion(String apiVersion) {
```
Do we need this method? Not sure where it is being used.

rwackes replied:
Yes. Ended up using it in second PR when building out the Perspective API call to set the api version to "v1alpha1". In future, more versions of this API may come out as they improve their models and make any changes to how they score/normalize scores.

vkkpng-g commented on capstone/src/main/java/com/google/sps/perspective/response/AnalyzeCommentResponse.java:
```diff
@@ -0,0 +1,67 @@
+   * Returns a mapping of Attribute names to the summary score as a probability
+   * @return a mapping of Attribute names to the summary score as a probability
+   */
+  public Map<String, Float> getAttributeSummaryScores() {
```
Is this part the most updated version of how u are calculating the summary score?

rwackes replied:
This is how I returned a Map of all the attribute summary scores that were calculated by the Perspective API and returned as part of the response to my java servlet. I use this map (in more recent PR) to check if any value is above our threshold of "flagging" and to transform each summary score into the format we want.

vkkpng-g commented on capstone/src/main/webapp/results.html:
```diff
@@ -1,14 +1,55 @@
 <!DOCTYPE html>
-<html>
+<html lang="en">
```
Is this html still being used for the final page. I know that u both did a redesign of the results page.

rwackes replied:
While the navbar html remains the same, the main content has obviously changed in later PRs that are dedicated to updating the UI of the results webpage. Cassandra has PRs for updating her image effect UI and I have a PR for updating my audio effect UI. This PR is not for UI

rwackes commented on capstone/src/main/java/com/google/sps/data/Transcribe.java:
```diff
@@ -0,0 +1,122 @@
+   * @return a string containing the audio transcription of the video file
+   */
+  public static String transcribeAudio() {
+    try {
```
I made that change in my second PR https://github.com/cassmarcussen/video-vigilance/pull/8/files#diff-2720d161695e54db71f42ea3aec1b94b . I'll copy here as well.

rwackes replied:
Done

rwackes commented on capstone/src/main/java/com/google/sps/data/Transcribe.java:
```diff
@@ -0,0 +1,122 @@
+            SpeechRecognitionAlternative alternative = speechTranscription.getAlternatives(0);
+            transcription = transcription + alternative.getTranscript();
+          } else {
+            System.out.println("No transcription found");
```
Done

rwackes commented on capstone/src/main/java/com/google/sps/data/Transcribe.java:
```diff
@@ -0,0 +1,122 @@
+        client.annotateVideoAsync(request);
+
+      // Wait for the video to be processed/for above operation to be complete.
+      AnnotateVideoResponse response = future.get(600, TimeUnit.SECONDS);
```
future.get(long timeout, TimeUnit unit) waits if necessary for at most the given time for the computation to complete, and then retrieves its result, if available. In this case, it waits at most 600 seconds (because that's what VI API uses in their example code) for the above line (OperationFuture<AnnotateVideoResponse, ...) to complete before retrieving the result

rwackes commented on capstone/src/main/java/com/google/sps/data/Transcribe.java:
```diff
@@ -0,0 +1,122 @@
+      AnnotateVideoResponse response = future.get(600, TimeUnit.SECONDS);
+
+      // Retrieve the first result since only one video was processed.
+      VideoAnnotationResults result = response.getAnnotationResults(0);
```
No. If there's no video to process, it would throw an exception before getting to this line.

rwackes commented on capstone/src/main/java/com/google/sps/data/Transcribe.java:
```diff
@@ -0,0 +1,122 @@
+            System.out.println("No transcription found");
+            transcription = "Hardcoded message. If this returns, that means there was no transcription found.";
+          }
+        } catch (IndexOutOfBoundsException ioe) {
```
Yes but the IndexOutOfBoundsException isnt for speechTranscription.getAlternatives() it's for alternatives.getTranscription() or alternatives.getConfidence() for the case where VI could not retrieve audio for that section of the video and thus there is no transcription or confidence annotation inside that section to retrieve.

rwackes commented on capstone/src/main/java/com/google/sps/perspective/PerspectiveAPI.java:
```diff
@@ -0,0 +1,43 @@
+   * Create AnalyzeCommentRequest with built out instance.
+   * @return the response of the Perspective API AnalyzeCommentRequest
+   */
+  public AnalyzeCommentRequest analyze() {
```
I renamed to createRequest() to better reflect the specific behavior of this method. Note: this method creates an AnalyzeCommentRequest for which we set the comment, attributes, language for the request we want to send and (in our Java servlet) we call postAsync() which then triggers it to be analyzed.
